File: problem2/problem2.erl
% Problem link: https://projecteuler.net/problem=2


% Problem statement:
    
%     By considering the terms in the Fibonacci sequence whose values do not 
%     exceed four million, find the sum of the even-valued terms.

-module(problem2).
-export([main/1]).

main( Max_value ) 
->
    Answer = sum_even_fibonacci( Max_value ),
    Answer
.

sum_even_fibonacci( Max_value ) 
->
    Index = 0,
    Acc = 0,
    F_t = 1,
    F_tminus1 = 1,
    sum_even_fibonacci( Index, 
                        Acc, 
                        F_t, 
                        F_tminus1, 
                        Max_value )
.

sum_even_fibonacci( _Index, 
                    Acc, 
                    F_t, 
                    _F_tminus1, 
                    Max_value ) when F_t > Max_value 
->
    Acc
;
sum_even_fibonacci( Index, 
                    Acc, 
                    F_t, 
                    F_tminus1, 
                    Max_value ) when Index rem 3 == 0 
->
    sum_even_fibonacci( Index + 1, 
                        Acc + F_t + F_tminus1, 
                        F_t + F_tminus1, 
                        F_t, 
                        Max_value )
;
sum_even_fibonacci( Index, 
                    Acc, 
                    F_t, 
                    F_tminus1, 
                    Max_value ) 
->
    sum_even_fibonacci( Index + 1, 
                        Acc, 
                        F_t + F_tminus1, 
                        F_t, 
                        Max_value )
.

File: problem2/problem2.c
// Problem link: https://projecteuler.net/problem=2

/*
Problem statement:
    
    By considering the terms in the Fibonacci sequence whose values do not 
    exceed four million, find the sum of the even-valued terms.

*/

#include <stdio.h>
#define MAX_VALUE 4000000

long long SumEvenFibonacci(long);

int main(int argc, char **argv)
{
    long long answer = SumEvenFibonacci(MAX_VALUE);

    printf("Answer: %lli\n", answer);
    return(0);
}

long long SumEvenFibonacci(long max_value)
{
    long long answer = 0;
    long long f_t = 1;
    long long f_t_minus_1 = 1;
    long long temp;
    for(int i = 0; f_t < max_value; i++)
    {
        temp = f_t + f_t_minus_1;
        f_t_minus_1 = f_t;
        f_t = temp;
        if(i % 3 == 0)
        {
            answer += temp;
        }
    }
    return answer;
}